src/main/java/eu/midnightdust/lib/config/MidnightConfig.java

```diff
@@ -0,0 +1,414 @@
+
+  public static void init(String modid, Class<?> config) {
+    path = FabricLoader.getInstance().getConfigDir().resolve(modid + ".json");
+    configClass.put(modid, config);
+
+    for (Field field : config.getFields()) {
+      EntryInfo info = new EntryInfo();
+      if (field.isAnnotationPresent(Entry.class) || field.isAnnotationPresent(Comment.class))
+        if (FabricLoader.getInstance().getEnvironmentType() == EnvType.CLIENT) initClient(modid, field, info);
+      if (field.isAnnotationPresent(Entry.class))
+        try {
+          info.defaultValue = field.get(null);
+        } catch (IllegalAccessException ignored) {}
+    }
+    try { gson.fromJson(Files.newBufferedReader(path), config); }
+    catch (Exception e) { write(modid); }
+
+    for (EntryInfo info : entries) {
+      if (info.field.isAnnotationPresent(Entry.class))
+        try {
+          info.value = info.field.get(null);
+          info.tempValue = info.value.toString();
+        } catch (IllegalAccessException ignored) {
+        }
+    }
+  }
+  @Environment(EnvType.CLIENT)
+  private static void initClient(String modid, Field field, EntryInfo info) {
+    Class<?> type = field.getType();
+    Entry e = field.getAnnotation(Entry.class);
+    info.width = e != null ? e.width() : 0;
+    info.field = field;
+    info.id = modid;
+
+    if (e != null) {
+      if (type == int.class) textField(info, Integer::parseInt, INTEGER_ONLY, e.min(), e.max(), true);
+      else if (type == double.class) textField(info, Double::parseDouble, DECIMAL_ONLY, e.min(), e.max(), false);
+      else if (type == String.class) {
+        info.max = e.max() == Double.MAX_VALUE ? Integer.MAX_VALUE : (int) e.max();
+        textField(info, String::length, null, Math.min(e.min(), 0), Math.max(e.max(), 1), true);
+      } else if (type == boolean.class) {
+        Function<Object, Text> func = value -> new LiteralText((Boolean) value ? "True" : "False").formatted((Boolean) value ? Formatting.GREEN : Formatting.RED);
+        info.widget = new AbstractMap.SimpleEntry<ButtonWidget.PressAction, Function<Object, Text>>(button -> {
+          info.value = !(Boolean) info.value;
+          button.setMessage(func.apply(info.value));
+        }, func);
+      } else if (type.isEnum()) {
+        List<?> values = Arrays.asList(field.getType().getEnumConstants());
+        Function<Object, Text> func = value -> new TranslatableText(modid + ".midnightconfig." + "enum." + type.getSimpleName() + "." + info.value.toString());
+        info.widget = new AbstractMap.SimpleEntry<ButtonWidget.PressAction, Function<Object, Text>>(button -> {
+          int index = values.indexOf(info.value) + 1;
+          info.value = values.get(index >= values.size() ? 0 : index);
+          button.setMessage(func.apply(info.value));
+        }, func);
+      }
+    }
+    entries.add(info);
+  }
```

Comment:
what's all of this reflection for?

Reply:
why do you need reflection*

Reply:
I dunno, I didn't write it, it's copied from [here](https://github.com/TeamMidnightDust/MidnightLib/blob/main/src/main/java/eu/midnightdust/lib/config/MidnightConfig.java) which is fine thanks to MIT

Reply:
that'll be very slow. i'm sure there's a better way to do it